src/manifolds/Hyperrectangle.jl

```diff
@@ -0,0 +1,526 @@
+@doc raw"""
+    Hyperrectangle{T} <: AbstractManifold{ℝ}
+
+Hyperrectangle, also known as orthotope or box. This is a manifold with boundary with the
+standard Euclidean metric.
+
+# Constructor
+
+    Hyperrectangle(lb::AbstractArray, ub::AbstractArray)
+
+Generate the hyperrectangle of arrays such that each element of the array is between lower
+and upper bound with the same index.
+"""
+struct Hyperrectangle{T<:AbstractArray} <: AbstractDecoratorManifold{ℝ}
+    lb::T
+    ub::T
+    function Hyperrectangle(lb::T, ub::T) where {T<:AbstractArray}
+        for i in eachindex(lb, ub)
+            if lb[i] > ub[i]
+                throw(
+                    ArgumentError(
+                        "Lower bound at index $i ($(lb[i])) is greater than the upper bound ($(ub[i]))",
+                    ),
+                )
+            end
+        end
+        return new{T}(lb, ub)
+    end
+end
+
+function active_traits(f, ::Hyperrectangle, args...)
+    return merge_traits(
+        IsDefaultMetric(EuclideanMetric()),
+        IsDefaultConnection(LeviCivitaConnection()),
+    )
+end
+
+function check_point(M::Hyperrectangle, p)
+    if !(eltype(p) <: Real)
```

Comment:
Would it maybe be easier to check that p is on `Euclidean(length(M.lb))` ? or `size(lb)`?

Reply:
Sure, though re-using that check here still requires a non-trivial amount of code for constructing the manifold and parsing the result, so the complexity isn't reduced here much, if at all.

Reply:
If you feel its not worth it, we can also duplicate the code here.